### ISSUE TITLE
dataclass subclass incorrectly inheriting default fields (false negative)

### ISSUE THREAD

**Issue by kwshi** [microsoft/pyright#7702]
**Describe the bug**

Per [current Python documentation on `dataclasses` and inheritance](https://docs.python.org/3/library/dataclasses.html#inheritance), if one dataclass inherits from another dataclass but re-specifies the same property names, then the subclass properties take precedence:

> Because the fields are in insertion order, derived classes override base classes.
> ```python
> @dataclass
> class Base:
>     x: Any = 15.0
>     y: int = 0
>
> @dataclass
> class C(Base):
>     z: int = 10
>     x: int = 15
> ```
> The final list of fields is, in order, `x`, `y`, `z`. The final type of `x` is `int`, as specified in class `C`.

By default, dataclasses are invariant in their property types, meaning that subclasses _may not_ override properties from the base classes with different types. So technically this example is unsound. The reason for invariance is mutability, so if instead the dataclasses were constructed with `frozen=True`, then they become _covariant_ in their property types, allowing subclasses to replace property types with more restrictive ones.

This much is OK. Problems arise when properties on the base class are given **default values**-- those default values can get incorrectly inherited by the sub-classes. This is ultimately a type soundness issue in Python's current execution semantics and arguably not a Pyright _bug_ per se. However, it does allow certain runtime `TypeError`s to slip past the type-checker completely undetected, which is why I'm flagging this as a Pyright "bug".

**Code or Screenshots**

```python
# pyright: strict
import dataclasses as dc

@dc.dataclass(frozen=True)
class Base:
    foo: int | None = dc.field(default=None)

@dc.dataclass(frozen=True)
class Sub(Base):
    foo: int

x = Sub()
# pyright flags no error here, which is consistent with Python's current runtime behavior, but this is unsound.
# interestingly enough, mypy manages to flag this as an error.

print(x.foo) # `None`, despite type inference currently believing that `x.foo` has type `int`
print(x.foo + 2) # type error! cannot add `None` to `int`
```

**VS Code extension or command-line**
1.1.354 CLI & nvim LSP

**Comment by erictraut** [comment]
Thanks for pointing this out.

I think mypy's error is incorrect here — or at the very least, misleading. Mypy also emits an error in the case where there is a valid (correctly-typed) default provided by the base class.

```python
from dataclasses import dataclass

@dataclass(frozen=True)
class Base:
    foo: int = 0

@dataclass(frozen=True)
class Sub(Base):
    foo: int

x = Sub() # Mypy says that `foo` is missing a positional argument
```

Mypy also doesn't correctly handle the case where a kw-only dataclass overrides a field in a non-kw-only dataclass base class.

```python
@dataclass(frozen=True)
class Base:
    foo: int = 0

@dataclass(frozen=True, kw_only=True)
class Sub(Base):
    foo: int = 0

x = Sub()
```

Maybe the best thing to do here is to require that if a base class provides a default value for a field and a subclass overrides that field, it must also provide a default value. This is effectively what mypy appears to be checking for, but it's strangely reporting the error at the time the dataclass constructor is called rather that within the class definition.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.359, which I just published.
